pkg/template/event.go

```diff
@@ -169,26 +169,65 @@ func getHeaderValue(header map[string][]string, headerName string) (string, erro
 func ResolveParams(bindings []*triggersv1.TriggerBinding, body []byte, header map[string][]string, params []pipelinev1.ParamSpec) ([]pipelinev1.Param, error) {
 	out, err := MergeBindingParams(bindings)
 	if err != nil {
-		return nil, xerrors.Errorf("error merging trigger params: %v", err)
+		return nil, xerrors.Errorf("error merging trigger params: %q", err)
 	}
-	out, err = ApplyBodyToParams(body, out)
+	event, err := NewEvent(body, header)
 	if err != nil {
-		return nil, xerrors.Errorf("error applying body to trigger params: %s", err)
+		return nil, xerrors.Errorf("failed to create Event: %q", err)
 	}
-	out, err = ApplyHeaderToParams(header, out)
+	out, err = ApplyEventValuesToParams(event, out)
 	if err != nil {
-		return nil, xerrors.Errorf("error applying header to trigger params: %s", err)
+		return nil, xerrors.Errorf("failed to ApplyEventValuesToParams: %q", err)
 	}
-
 	return MergeInDefaultParams(out, params), nil
 }
 
+// ResolveResources resolves a templated resource by replacing params with their values.
 func ResolveResources(template *triggersv1.TriggerTemplate, params []pipelinev1.Param) []json.RawMessage {
 	resources := make([]json.RawMessage, len(template.Spec.ResourceTemplates))
-	uid := UID()
 	for i := range template.Spec.ResourceTemplates {
 		resources[i] = ApplyParamsToResourceTemplate(params, template.Spec.ResourceTemplates[i].RawMessage)
-		resources[i] = ApplyUIDToResourceTemplate(resources[i], uid)
+		resources[i] = ApplyUIDToResourceTemplate(resources[i], UID())
 	}
 	return resources
 }
+
+
+// Event represents a HTTP event that Triggers processes
+type Event struct {
+	Header map[string]string `json:"header"`
+	Body   interface{}       `json:"body"`
+}
+
+// NewEvent returns a new Event from HTTP headers and body
+func NewEvent(body []byte, headers map[string][]string) (Event, error) {
+	var data interface{}
+	if len(body) > 0 {
+		if err := json.Unmarshal(body, &data); err != nil {
+			return Event{}, xerrors.Errorf("failed to unmarshal request body: %q", err)
+		}
+	}
+	joinedHeaders := make(map[string]string, len(headers))
+	for k, v := range headers {
+		joinedHeaders[k] = strings.Join(v, ",")
+	}
+
+	return Event{
+		Header: joinedHeaders,
+		Body:   data,
+	}, nil
+}
+
+
+// ApplyEventValuesToParams returns a slice of Params with the JSONPath variables replaced
+// with values from the event body and headers.
+func ApplyEventValuesToParams(ec Event, params []pipelinev1.Param) ([]pipelinev1.Param, error) {
```

Comment:
super nit: my preference is for this to be (params []pipelinev1.Param, ec Event), since the params are the primary object we are acting on here.

Reply:
done!